File: Sprint8/E/C++/main.cpp
// https://contest.yandex.ru/contest/26131/problems/E/
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

int main() {
    std::string s;
    std::cin >> s;

    int n;
    std::cin >> n;

    std::vector<std::pair<std::string, int>> strs(n);
    for (int i = 0; i < n; ++i) {
        std::string t;
        int k;
        std::cin >> t >> k;
        strs[i] = {t, k};
    }

    std::sort(strs.begin(), strs.end(), [](std::pair<std::string, int> a, std::pair<std::string, int> b) {return a.second < b.second;});

    std::string result;
    int prev = 0;
    for (auto &str : strs) {
        int pos = str.second;
        result += s.substr(prev, pos - prev) + str.first;
        prev = pos;
    }

    result += s.substr(prev);

    std::cout << result << std::endl;
}

File: Sprint8/E/Rust/src/main.rs
// https://contest.yandex.ru/contest/26131/problems/E/
use std::io::BufRead;

struct Replacement {
    value: String,
    position: usize,
}

fn read_replacements(lines: &mut std::io::Lines<std::io::StdinLock>, n: usize) -> Vec<Replacement> {
    let mut replacements = Vec::with_capacity(n);
    for _ in 0..n {
        let line = lines.next().unwrap().unwrap();
        let parts: Vec<&str> = line.split_whitespace().collect();
        let replacement = Replacement {
            value: parts[0].to_string(),
            position: parts[1].parse().unwrap(),
        };
        replacements.push(replacement);
    }
    replacements.sort_by(|a, b| a.position.cmp(&b.position));
    replacements
}

fn apply_replacements(s: &str, replacements: &[Replacement]) -> String {
    let mut result = String::new();
    let mut prev = 0;
    for r in replacements {
        result.push_str(&s[prev..r.position]);
        result.push_str(&r.value);
        prev = r.position;
    }
    result.push_str(&s[prev..]);
    result
}

fn main() {
    let mut lines = std::io::stdin().lock().lines();
    let s = lines.next().unwrap().unwrap();
    let n: usize = lines.next().unwrap().unwrap().parse().unwrap();

    let replacements = read_replacements(&mut lines, n);
    println!("{}", apply_replacements(&s, &replacements));
}
